### PR TITLE
util: Revert back `MoveFileExW` call for MinGW-w64

### PR DESCRIPTION
**Pull request by hebasto**
Unfortunately, bitcoin/bitcoin#24308 introduced a [regression](https://github.com/bitcoin/bitcoin/pull/24308#issuecomment-1037259386) for mingw builds.

The root of the problem is a broken implementation of [`std::filesystem::rename`](https://en.cppreference.com/w/cpp/filesystem/rename). In particular, the expected behavior
> If `old_p` is a non-directory file, then `new_p` must be ... existing non-directory file: `new_p` _is first deleted_...

fails with the "File exists" error.

This PR reverts back the `MoveFileExW` call, and adds the [suggested](https://github.com/bitcoin/bitcoin/pull/24308#pullrequestreview-878832906) unit test.

### CODE REVIEW

**Comment by fanquake** [comment]
> The root of the problem is a broken implementation of [std::filesystem::rename](https://en.cppreference.com/w/cpp/filesystem/rename).

What version is broken? Is there an upstream issue / bug report?

**Comment by MarcoFalke** [comment]
Ah, it fails: https://cirrus-ci.com/task/6670169794674688?logs=ci#L4536

Concept ACK

**Comment by laanwj** [comment]
Good catch… Leave it to windows libraries to not actually implement the standard as it is defined.

**Comment by MarcoFalke** [comment]
At one point it might be better to drop support for mingw and use clang instead for cross compiling. With all the bugs we are seeing and other projects moving to clang I am not convinced this is a good long term choice anymore.

File: src/util/system.cpp [hunk 1]
```diff
@@ -1062,9 +1062,14 @@ void ArgsManager::LogArgs() const
 
 bool RenameOver(fs::path src, fs::path dest)
 {
+#ifdef __MINGW64__
+    return MoveFileExW(src.wstring().c_str(), dest.wstring().c_str(),
```
Comment (laanwj) [inline src/util/system.cpp:1066]: Might want to add a comment here specifying why this workaround exists, and we should probably file an upstream issue and refer to it, so it has a limited lifetime.

**Comment by laanwj** [comment]
I didn't realize the bug was Mingw only. So MSVC (or even clang) implements this correctly?

> At one point it might be better to drop support for mingw and use clang instead for cross compiling.

Agree. If it's less buggy than mingw-w64, I think it would be good to consider using clang for the windows build.

File: src/test/fs_tests.cpp [hunk 2]
```diff
@@ -118,4 +118,37 @@ BOOST_AUTO_TEST_CASE(fsbridge_fstream)
...
+    BOOST_CHECK(RenameOver(path1, path2));
+
+    BOOST_CHECK(!fs::exists(path1));
+
+    {
+        std::ifstream file{path2};
+        std::string contents;
+        file >> contents;
+        BOOST_CHECK_EQUAL(contents, path1_contents);
+    }
```
Comment (vasild) [inline src/test/fs_tests.cpp:151]: It is a good practice to clean up after the test, even if it is a temporary folder. I forgot that, but woke up during the night with the thought "file cleanup was forgotten!".

```suggestion
    }
    fs::remove(path2);
```

**Review by vasild — APPROVED** [review]
ACK 15a043dcee116f5273a262ad83853c791c50b9fa

**Comment by vasild** [comment]
mingw32 works ok? Or we don't support that?

**Comment by hebasto** [comment]
> mingw32 works ok? Or we don't support that?

The latter, I [guess](https://github.com/bitcoin/bitcoin/pull/15939).

**Comment by hebasto** [comment]
Updated 15a043dcee116f5273a262ad83853c791c50b9fa -> 4d92107041c508cf7bca47dce088979e519fb8bc ([pr24331.01](https://github.com/hebasto/bitcoin/commits/pr24331.01) -> [pr24331.02](https://github.com/hebasto/bitcoin/commits/pr24331.02), [diff](https://github.com/hebasto/bitcoin/compare/pr24331.01..pr24331.02)):

- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/24331#discussion_r805652748)

File: src/test/fs_tests.cpp (hunk 2 above)
Comment (hebasto) [reply src/test/fs_tests.cpp:151]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24331#issuecomment-1038872961).

**Review by vasild — APPROVED** [review]
ACK 4d92107041c508cf7bca47dce088979e519fb8bc

**Comment by fanquake** [comment]
> What version is broken? Is there an upstream issue / bug report?

**Comment by laanwj** [comment]
> mingw32 works ok? Or we don't support that?

afaik mingw-w64 was a fork years ago because mingw32 wasn't really maintained anymore, i doubt they support newer C++ standards required

**Comment by hebasto** [comment]
Updated 4d92107041c508cf7bca47dce088979e519fb8bc -> dc01cbc538765f64326bca30952c83e3862d0d54 ([pr24331.02](https://github.com/hebasto/bitcoin/commits/pr24331.02) -> [pr24331.03](https://github.com/hebasto/bitcoin/commits/pr24331.03), [diff](https://github.com/hebasto/bitcoin/compare/pr24331.02..pr24331.03)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/24331#discussion_r805632060)

File: src/util/system.cpp (hunk 1 above)
Comment (hebasto) [reply src/util/system.cpp:1072]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24331#issuecomment-1039185324).

**Review by vasild — APPROVED** [review]
ACK dc01cbc538765f64326bca30952c83e3862d0d54

**Comment by fanquake** [comment]
> What version(s) are broken? Is there an upstream issue / bug report?

This is still unclear to me?

**Comment by laanwj** [comment]
> This is still unclear to me?

FWIW there's a comment in the source
```
    // This bug has been fixed in upstream:
    //  - GCC 10.3: 8dd1c1085587c9f8a21bb5e588dfe1e8cdbba79e
    //  - GCC 11.1: 1dfd95f0a0ca1d9e6cbc00e6cbfd1fa20a98f312
```

**Pull request merged**